### PR TITLE
Add BarChart component

### PR DESCRIPTION
**Pull request by samshara**

### CODE REVIEW

**Review by tnagorra — COMMENTED** [review]
I am not sure we are merging this to toggle-ui. Do we need this component?

File: src/components/BarChart/index.tsx [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        colorSelector,
+        margin: {
+            top,
+            right,
+            bottom,
+            left,
+        },
+    }  = props;
+
+    const svgRef = useRef(null);
```
Comment (tnagorra) [inline src/components/BarChart/index.tsx]: Type should be defined for svgRef

File: src/components/BarChart/index.tsx [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+            .append('g')
+            .attr('fill', (d: K) => colorSelector(d))
+            .selectAll('.bar')
+            .data(data)
+            .join('rect')
+            .attr('x', (d: K) => x(labelSelector(d)))
+            .attr('y', (d: K) => y(valueSelector(d)))
+            .attr('height', (d: K) => y(0) - y(valueSelector(d)))
+            .attr('width', x.bandwidth());
+    }, [data]);
```
Comment (tnagorra) [inline src/components/BarChart/index.tsx]: Not all dependencies are listed here!

File: src/components/BarChart/index.tsx [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+            .padding(0.1);
+
+        const y = scaleLinear()
+            .domain([0, max(data, valueSelector)])
+            .nice()
+            .range([height - bottom, top]);
+
+        const xAxis = axisBottom(x);
+
+        const yAxis = axisLeft(y);
```
Comment (tnagorra) [inline src/components/BarChart/index.tsx]: can we not just memoize these?

Comment (tnagorra) [reply src/components/BarChart/index.tsx]: i don't know if it's better to use separate useEffect hooks for different side-effects (d3 invocation)

File: src/components/BarChart/index.tsx (hunk 2 above)
Comment (frozenhelium) [reply src/components/BarChart/index.tsx]: This is not the d4 method. @tnagorra should we use d4 method or this looks good?

Comment (tnagorra) [reply src/components/BarChart/index.tsx]: I don't think this is how we should write them!
Can you explain a bit more on the d4 method though?

Comment (tnagorra) [reply src/components/BarChart/index.tsx]: Without select, selectAll, enter, append, data, attr, style?

Comment (tnagorra) [reply src/components/BarChart/index.tsx]: @samshara for reference, https://d4.js.org/

File: src/components/BarChart/index.tsx [hunk 4]
```diff
@@ -38,61 +35,108 @@ function BarChart<K>(props: Props<K>) {
...
+        .padding(0.1),
+    [data, labelSelector, left, right, width]);
+
+    const y = useMemo(() => scaleLinear()
+        .domain([0, max(data, valueSelector)])
+        .nice()
+        .range([height - bottom, top]),
+    [data, valueSelector, bottom, top, height]);
+
+    const bars = useMemo(() => data.map((d) => (
```
Comment (tnagorra) [inline src/components/BarChart/index.tsx]: Maybe we don't need to memoize the jsx outputs?

**Review by tnagorra — APPROVED** [review]

File: src/components/BarChart/index.tsx [hunk 5]
```diff
@@ -0,0 +1,152 @@
+import React, { useMemo } from 'react';
+
+import { scaleBand, scaleLinear } from 'd3-scale';
+import { max } from 'd3-array';
```
Comment (frozenhelium) [inline src/components/BarChart/index.tsx]: @tnagorra we should use `max` from fujs right?

File: package.json [hunk 6]
```diff
@@ -42,6 +42,10 @@
         "@babel/runtime-corejs3": "^7.11.2",
         "@storybook/client-api": "^6.0.21",
         "@togglecorp/fujs": "^1.9.2-beta.0",
+        "d3-array": "^2.8.0",
```
Comment (frozenhelium) [inline package.json]: is this required?

File: package.json [hunk 7]
```diff
@@ -42,6 +42,10 @@
...
+        "d3-axis": "^2.0.0",
+        "d3-scale": "^3.2.3",
+        "d3-selection": "^2.0.0",
```
Comment (frozenhelium) [inline package.json]: is this required?

**Pull request merged**